lib/raven/base.rb

```diff
@@ -14,12 +14,49 @@
 require 'raven/processor/sanitizedata'
 require 'raven/processor/removecircularreferences'
 require 'raven/processor/utf8conversion'
+require 'binding_of_caller'
 
 major, minor, patch = RUBY_VERSION.split('.').map(&:to_i)
 if (major == 1 && minor < 9) || (major == 1 && minor == 9 && patch < 2)
   require 'raven/backports/uri'
 end
 
+module Kernel
+  def self.monkey_patch_raise
+    if @@monkey_patch_raise_occur == true
+      return
+    end
+    apply_monkey_patch_raise
+    @@monkey_patch_raise_occur = true
+  end
+
+  private
+  def self.apply_monkey_patch_raise
+    original_raise = method(:raise)
+    define_method(:raise) do |*args|
+      begin
+        original_raise.call(*args)
+      rescue Exception => e
+        prev_stack_info = e.instance_variable_get(:@stack_info)
+        if prev_stack_info.nil?
+          pop_length = 2
+
+          callers = binding.callers.drop(pop_length)
+          e.instance_variable_set(:@stack_info, callers)
+          e.set_backtrace(e.backtrace.drop(pop_length))
+
+          callers.each_with_index do |caller_obj, idx|
+            e.backtrace[idx].instance_variable_set(:@stack_info, caller_obj)
+          end
+        end
+        original_raise.call(e)
+      end
+    end
+  end
+
+  @@monkey_patch_raise_occur = false
```

Comment:
1. add @@monkey_patch_raise_mutex.synchronize { ... } 
2. check flag twice (outside of critical section, inside of critical section)